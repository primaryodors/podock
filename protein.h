
#include "aminoacid.h"

#ifndef _PROTEIN
#define _PROTEIN

#include <string>

void ext_mtl_coord_cnf_cb(int iter);

class Region
{
public:
    int start=0;
    int end=0;
    std::string name="";
};

class Protein
{
public:
    // Constructors.
    Protein(const char* name);

    // Build functions.
    bool add_residue(const int resno, const char aaletter);
    bool add_residue(const char* pdbdata);
    bool add_sequence(const char* sequence);
    float coordinate_metal(const char* metal_elemsym, const int* coord_residues);		// Returns coordination anomaly.
    void set_clashables();
    void delete_residue(int resno);
    void delete_sidechain(int resno);
    void delete_residues(int startres, int endres);
    void delete_sidechains(int startres, int endres);
    MetalCoord* coordinate_metal(Atom* metal, int residues, int* resnos, char** res_anames);
    void set_region(std::string name, int start, int end);

    // Serialization.
    int load_pdb(FILE* instream);				// Returns number of residues loaded.
    void save_pdb(FILE* outstream);
    void end_pdb(FILE* outstream);

    // Getters.
    int get_seq_length();
    int get_start_resno();
    AminoAcid* get_residue(int resno);
    Molecule* metals_as_molecule();
    AminoAcid** get_residues_can_clash(int resno);
    bool aa_ptr_in_range(AminoAcid* aaptr);
    Region get_region(std::string name);
    int get_region_start(std::string name);
    int get_region_end(std::string name);
    Atom* get_atom(int resno, const char* aname);
    Point get_atom_location(int resno, const char* aname);

    // Metrics functions.
    float get_internal_clashes();
    float get_intermol_clashes(const Molecule* ligand);
    float get_intermol_binding(const Molecule* ligand);
    int get_residues_can_clash_ligand
    (	AminoAcid** reaches_spheroid,
        const Molecule* ligand,
        const Point nodecen,
        const Point size,
        const int* mcoord_resno
    );
    float get_helix_orientation(int startres, int endres);

    // Motion functions
    void rotate_backbone(int residue_no, bb_rot_dir direction, float angle);
    void rotate_backbone_partial(int startres, int endres, bb_rot_dir direction, float angle);
    void conform_backbone(int startres, int endres, int iters = 50, bool backbone_atoms_only = false);
    void conform_backbone(int startres, int endres, Atom* a, Point target, int iters = 50);
    void conform_backbone(int startres, int endres, Atom* a1, Point target1, Atom* a2, Point target2, int iters = 50);
    void make_helix(int startres, int endres, float phi, float psi);
    void make_helix(int startres, int endres, int stopat, float phi, float psi);
    float orient_helix
    (	int startres, int endres,						// Boundaries of helix.
        int stopat,										// Last residue to move with helix.
        float angle,									// 0 = horizontal; positive = ascending (+Y) with increasing resno.
        int iterations
    );

    int* mcoord_resnos = NULL;

protected:
    std::string name;
    char* sequence=0;
    AminoAcid** residues=0;
    AminoAcid*** res_can_clash = 0;
    Atom** ca=0;
    float* res_reach=0;
    Atom** metals=0;
    int metcount=0;
    Star aaptrmin, aaptrmax;
    MetalCoord** m_mcoord=0;
    Region regions[PROT_MAX_RGN] {};

    int* get_residues_in_reach(int resno);
    float get_coord_anomaly(Atom* metal, AminoAcid* coord_res);
    void conform_backbone(int startres, int endres,
                          Atom* a1, Point target1,
                          Atom* a2, Point target2,
                          int iters, bool backbone_atoms_only
                         );
    void mtl_coord_cnf_cb(int iter);
    friend void ext_mtl_coord_cnf_cb(int iter);
};


#endif

