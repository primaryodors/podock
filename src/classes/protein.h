
#include "aminoacid.h"

#ifndef _PROTEIN
#define _PROTEIN

#include <string>

void ext_mtl_coord_cnf_cb(int iter);

struct Region
{
    int start=0;
    int end=0;
    std::string name="";
};

class Protein
{
	public:
    // Constructors.
    Protein(const char* name);

    // Build functions.
    bool add_residue(const int resno, const char aaletter);
    bool add_residue(const char* pdbdata);
    bool add_sequence(const char* sequence);
    float coordinate_metal(const char* metal_elemsym, const int* coord_residues);		// Returns coordination anomaly.
    void set_clashables();
    void delete_residue(int resno);
    void delete_sidechain(int resno);
    void delete_residues(int startres, int endres);
    void delete_sidechains(int startres, int endres);
    MetalCoord* coordinate_metal(Atom* metal, int residues, int* resnos, char** res_anames);
    void set_region(std::string name, int start, int end);

    // Serialization.
    int load_pdb(FILE* infile);				// Returns number of residues loaded.
    void save_pdb(FILE* outfile);
    void end_pdb(FILE* outfile);

    // Getters.
    int get_seq_length();
    std::string get_sequence();
    int get_start_resno();
    AminoAcid* get_residue(int resno);
    Molecule* metals_as_molecule();
    bool aa_ptr_in_range(AminoAcid* aaptr);
    Region get_region(std::string name);
    int get_region_start(std::string name);
    int get_region_end(std::string name);
    Atom* get_atom(int resno, const char* aname);
    Point get_atom_location(int resno, const char* aname);
    std::vector<std::string> get_remarks(std::string search_for = "");
    std::string get_name() { return std::string(name); }

    // Metrics functions.
    float get_internal_clashes();
    float get_intermol_clashes(const Molecule* ligand);
    float get_intermol_binding(const Molecule* ligand);
    AminoAcid** get_residues_can_clash(int resno);
    int get_residues_can_clash_ligand
    (	AminoAcid** reaches_spheroid,
        const Molecule* ligand,
        const Point nodecen,
        const Point size,
        const int* mcoord_resno
    );
    float get_helix_orientation(int startres, int endres);
    Point get_region_center(int startres, int endres);

    // Motion functions
    void move_piece(int start_res, int end_res, Point new_center);		// After calling this, you should reconnect the broken ends with conform_backbone().
    void rotate_piece(int start_res, int end_res, int align_res, Point align_target, int pivot_res = 0);		// If no pivot res, rotate about the center.
    
    void rotate_backbone(int residue_no, bb_rot_dir direction, float angle);
    void rotate_backbone_partial(int startres, int endres, bb_rot_dir direction, float angle);
    void conform_backbone(int startres, int endres, int iters = 50, bool backbone_atoms_only = false);
    void conform_backbone(int startres, int endres, Atom* a, Point target, int iters = 50);
    void conform_backbone(int startres, int endres, Atom* a1, Point target1, Atom* a2, Point target2, int iters = 50);
    void make_helix(int startres, int endres, float phi, float psi);
    void make_helix(int startres, int endres, int stopat, float phi, float psi);
    float orient_helix
    (	int startres, int endres,						// Boundaries of helix.
        int stopat,										// Last residue to move with helix.
        float angle,									// 0 = horizontal; positive = ascending (+Y) with increasing resno.
        int iterations
    );

    int* mcoord_resnos = NULL;

	protected:
    std::string name;
    char* sequence=0;
    AminoAcid** residues=0;
    AminoAcid*** res_can_clash = 0;
    Atom** ca=0;
    float* res_reach=0;
    Atom** metals=0;
    int metcount=0;
    Star aaptrmin, aaptrmax;
    MetalCoord** m_mcoord=0;
    Region regions[PROT_MAX_RGN];
    std::vector<string> remarks;

    int* get_residues_in_reach(int resno);
    float get_coord_anomaly(Atom* metal, AminoAcid* coord_res);
    void conform_backbone(int startres, int endres,
                          Atom* a1, Point target1,
                          Atom* a2, Point target2,
                          int iters, bool backbone_atoms_only
                         );
    void mtl_coord_cnf_cb(int iter);
    friend void ext_mtl_coord_cnf_cb(int iter);
};


#endif

