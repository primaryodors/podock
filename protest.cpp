#include <cstring>
#include <iostream>
#include <stdio.h>
#include <string>
#include <math.h>
#include "protein.h"

using namespace std;

int main(int argc, char** argv)
{
    int i, seqarg=1, namearg=0;
    
    if (argc < 2)
    {
    	cout << "Usage:" << endl << "protest [sequence]" << endl;
    	cout << endl;
    	cout << "Optionally, you can override the default aminos.dat file with --dat [filename], for example if using non-standard amino acids.";
    	cout << endl;
    	return -1;
    }

    if (!strcmp("--dat", argv[seqarg]))
    {
        override_aminos_dat = argv[seqarg+1];
        seqarg += 2;
    }

    if (!strcmp("--name", argv[seqarg]))
    {
        namearg = seqarg+1;
        seqarg += 2;
    }

    Protein p(namearg?argv[namearg]:"Test");
    p.add_sequence(argv[seqarg]);
    
    // Stretch it out.
    Point pt = p.get_atom_location(1, "N");
    pt.x += 10000;
    int len = p.get_seq_length();
    p.conform_backbone(1, len, p.get_atom(len, "C"), pt, 50);

    const char* outfn = "test.pdb";
    FILE* pf = fopen(outfn, "wb");
    p.save_pdb(pf);
    p.end_pdb(pf);
    fclose(pf);
    cout << "Wrote " << outfn << endl;

    Molecule m("Test2");
    pf = fopen(outfn, "rb");
    m.from_pdb(pf);
    fclose(pf);
    // m.minimize_internal_collisions();

    const char* outfn2 = "test2.sdf";
    pf = fopen(outfn2, "wb");
    m.save_sdf(pf);
    fclose(pf);
    cout << "Wrote " << outfn2 << endl;

    Protein p1("Test3");
    pf = fopen(outfn, "rb");
    int rescount = p1.load_pdb(pf);
    fclose(pf);
    cout << "Read " << rescount << " residue(s)." << endl;


    return 0;
}













