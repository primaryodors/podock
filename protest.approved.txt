ARNDCEQGHILKMFPUSTWYV
Wrote test.pdb
Wrote test_alpha.pdb
Wrote test_beta.pdb
Wrote test_310.pdb
Wrote test_pi.pdb
Wrote test2.sdf
Read 21 residue(s).

